### PR TITLE
CNTRLPLANE-3727: Agentic context docs and contributing guide

### PR DESCRIPTION
**Pull request by bhperry**
## Summary by CodeRabbit

* **Documentation**
  * Added comprehensive contributor/agent guidance covering operator scope, contributor workflow rules, build/verify commands, and common do/don’t practices.
  * Added detailed architecture documentation describing the static pod lifecycle approach, TLS/certificate rotation flow, endpoints and health behavior, and testing strategy.
  * Added/expanded contribution guidelines covering coding and testing expectations plus pull request and verification workflows.
  * Centralized contributor instructions via a documentation link/symlink for easier discovery.

### CODE REVIEW

File: AGENTS.md [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+2. **Backup Credentials** - stored in secrets, never logged
+3. **Escape Hatches** - documented in docs/overview/overview.md (Secret Escape Hatches section)
+4. **Fencing Scripts** - in bindata/etcd/, reviewed for injection vulnerabilities
+
+## Key Files
+
+- `pkg/operator/starter.go` - Operator entrypoint, initializes all controllers
+- `pkg/operator/ceohelpers/bootstrap.go` - Bootstrap scaling logic
+- `pkg/operator/etcdendpointscontroller/` - Critical DNS-independent controller
+- `pkg/etcdcli/etcdclient.go` - etcd client abstraction
```
Comment (dusk125) [inline AGENTS.md]: `pkg/etcdcli/etcdcli.go` is the correct filename

File: AGENTS.md [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+│   ├── cluster-etcd-operator-tests-ext/  # OTE test binary
+│   ├── tnf-monitor/                      # Two Node Fencing monitor
+│   └── tnf-setup-runner/                 # Two Node Fencing setup
+├── docs/                # Documentation (HACKING, FAQ, overview)
+├── hack/                # Build and generation scripts
+├── jsonnet/             # Alert and dashboard definitions
+├── manifests/           # CVO-managed YAML (deployment, RBAC, monitoring)
+├── pkg/
+│   ├── operator/        # Core operator logic (20+ controllers)
+│   │   ├── etcdendpointscontroller/
```
Comment (dusk125) [inline AGENTS.md]: I dunno if it's worth enumerating all of the controllers here; having only a subset may mislead the bots and some of the omitted are important as well.

File: ARCHITECTURE.md [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+**Safety:** QuorumChecker blocks revisions if quorum unhealthy; bundle-rollout gating (4.17+) prevents cert updates before CA bundle rollout completes.
+
+## etcd Client Abstraction
+
+CEO's etcd client (`pkg/etcdcli/interfaces.go`) composes 10+ interfaces (MemberAdder, MemberRemover, Defragment, etc.). Two implementations:
+
+- **etcdcli.EtcdClient** - Direct client using TLS certs
+- **etcdcli.MemberCache** - Caching wrapper (60s TTL); not used for 2-node quorum checks
+
+**Connection fallback:** IP-based `openshift-etcd/endpoints` → DNS-based `openshift-etcd/etcd-endpoints` (with staleness detection) → direct node IPs.
```
Comment (dusk125) [inline ARCHITECTURE.md]: This isn't correct, it should be something like:
`Connection fallback: etcd-endpoints ConfigMap (IP-based, maintained by EtcdEndpointsController) → direct node internal IPs (from Node objects via GetPreferredInternalIPAddressForNodeName).`

File: ARCHITECTURE.md [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+See `pkg/operator/starter.go` `RunOperator()` for full list and initialization order.
+
+## Key Design Decisions
+
+### DNS Independence for EtcdEndpointsController
+
+**Decision:** The EtcdEndpointsController must never depend on DNS, directly or transitively.
+
+**Rationale:** This controller maintains `openshift-etcd/endpoints` ConfigMap with IP addresses that DNS resolution depends on. A DNS dependency would create a circular bootstrap problem.
```
Comment (dusk125) [inline ARCHITECTURE.md]: `openshift-etcd/etcd-endpoints` is the configmap name. If you want to be explicit about the namespace `openshift-etcd` and configmap name `etcd-endpoints`

**Review by dusk125 — CHANGES_REQUESTED** [review]

File: AGENTS.md (hunk 2 above)
Comment (bhperry) [reply AGENTS.md]: good point, will leave it at high level dir

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+We largely follow the [Kubernetes Code Conventions](https://github.com/kubernetes/community/blob/main/contributors/guide/coding-conventions.md#code-conventions).
+
+Review both the Kubernetes Code Conventions and the ones specified here.
+There will be some overlap. If any conventions are at odds with one another, prefer the conventions explicitly documented here.
+
+### Bash
+
+- Follow the [shell styleguide](https://google.github.io/styleguide/shellguide.html).
+- Use [`shellcheck`](https://github.com/koalaman/shellcheck) to identify common mistakes or caveats.
+- Ensure that all scripts run consistently across Linux and MacOS.
```
[the remark this answers is not in the document]
Comment (bhperry) [reply CONTRIBUTING.md]: MacOS is referrenced throughout the repo, leaving as is.

File: AGENTS.md [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+│   ├── cluster-etcd-operator/            # Main operator binary
+│   ├── cluster-etcd-operator-tests-ext/  # OTE test binary
+│   ├── tnf-monitor/                      # Two Node Fencing monitor
+│   └── tnf-setup-runner/                 # Two Node Fencing setup
+├── docs/                # Documentation (HACKING, FAQ, overview)
+├── hack/                # Build and generation scripts
+├── jsonnet/             # Alert and dashboard definitions
+├── manifests/           # CVO-managed YAML (deployment, RBAC, monitoring)
+├── pkg/
+│   ├── operator/        # Core operator logic (20+ controllers)
```
Comment (dusk125) [inline AGENTS.md:42]: Maybe bring back `ceohelpers` and controllers in the form `*controller`... although there are some other non-"controller" controllers ("bootstrapteardown", etc).

What do you think?

Comment (dusk125) [reply AGENTS.md:42]: Maybe leaving it as is and going into more complete detail in the Architecture file - going through all of the controllers?

Comment (bhperry) [reply AGENTS.md:42]: I think may not be necessary to explain ceohelpers. If agent lists the dir it's fairly self explanatory what that is.

Comment (bhperry) [reply AGENTS.md:42]: Enumerating the controllers with descriptions in ARCHITECTURE.md makes sense to me

Comment (dusk125) [reply AGENTS.md:42]: Both sound good to me!

File: AGENTS.md [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+        WithSync(syncer.Sync).
+        ToController("MyController", eventRecorder.WithComponentSuffix("my-controller"))
+}
+```
+
+Key elements:
+- Constructor function named `NewXxxController`
+- Wraps sync function with health checker
+- Uses library-go factory builder
+- Returns `factory.Controller` interface
```
Comment (tjungblu) [inline AGENTS.md:82]: a critical rule for me would be to double check whether all informers are covered (probably one of our 80% "bug" problem).

File: AGENTS.md [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+_, _, updateErr := v1helpers.UpdateStatus(ctx, c.operatorClient,
+    v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+        Type:    "MyControllerDegraded",
+        Status:  operatorv1.ConditionTrue,
+        Reason:  "SyncError",
+        Message: err.Error(),
+    }),
+)
+```
+
```
Comment (tjungblu) [inline AGENTS.md:99]: this is very tribal for this but useful for agents:
* use the degraded scenario very sparingly, only after a meaningful amount of time has passed (etcd moves slow due to the static pod machinery) and only if it really requires **actionable** human intervention. 
* avoid cases where the conditions can flap, always add an inertia component

File: AGENTS.md [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+### Bootstrap Scaling Strategies
+
+The operator enforces different quorum requirements during bootstrap:
+- `HAScalingStrategy` - requires 3 nodes (default)
+- `DelayedHAScalingStrategy` - allows 2 nodes during bootstrap, then requires 3
```
Comment (tjungblu) [inline AGENTS.md:110]: this could use a bit more context. IIRC that is for ass. installer where the bootstrap node is repurposed as the last master node. Meaning you can get away with only three nodes in total (especially useful in bare metal).

File: AGENTS.md [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+- `TwoNodeScalingStrategy` - requires 2 nodes (for Two Node OpenShift with Fencing deployments)
+- `DelayedTwoNodeScalingStrategy` - allows 1 node during bootstrap, then requires 2 (for Two Node OpenShift with Fencing deployments)
+- `BootstrapInPlaceStrategy` - single-node bootstrap where no separate bootstrap node exists (live ISO pivot)
+- `UnsafeScalingStrategy` - scaling without regard to node count or quorum
+
+Check `pkg/operator/ceohelpers/bootstrap.go` for implementation.
+
+## Important Constraints
+
+1. **DNS Independence** - The `EtcdEndpointsController` must never depend on DNS (directly or transitively) as it maintains the IP-based ConfigMap that DNS depends on
```
Comment (tjungblu) [inline AGENTS.md]: why is that: DNS has no guarantee that only a single IP and always the same is returned, hence it was decided to switch to IPs. 

Previously it happened that a member node thought it was a different member node (also flapping between IPs). Using only IPs is a core design choice in this operator and that very likely is not changing.

File: AGENTS.md [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+2. **CVO Resource Management** - Resources in `manifests/` are managed by CVO; removal requires special handling (see [CVO Resource Removal](#cvo-resource-removal))
```
Comment (tjungblu) [inline AGENTS.md]: it is also very important that the ordering is correctly reflected in the file names. See the recent OVN/networking problem where a manifest was deployed before the other and that caused outages in CI.

File: AGENTS.md [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+3. **Controller Removal** - Removing a controller requires cleaning up stale status conditions to prevent upgrade blockers
```
Comment (tjungblu) [inline AGENTS.md]: I would even say: an agent should never remove a controller. Besides the UpgradeBackup Controller we had, we didn't really have a removal in all those years.

File: AGENTS.md [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+4. **Static Pod Revisions** - Follow library-go static pod revision pattern; don't invent custom update logic
+5. **Namespace** - Primary namespace is `openshift-etcd`; target namespace constant is in `pkg/operator/operatorclient`
```
Comment (tjungblu) [inline AGENTS.md]:
```suggestion
5. **Namespace** - Primary namespace for the operator resources is `openshift-etcd-operator`; target namespace for the operand is `openshift-etcd`. Constants are found in `pkg/operator/operatorclient/interfaces.go` and should always be used 
```

File: AGENTS.md [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+
+Test files are in `test/e2e/` and registered via the OTE framework in `cmd/cluster-etcd-operator-tests-ext/`.
+
+## Alerts and Dashboards
+
+Prometheus alerts and Grafana dashboards are **generated from jsonnet**, not hand-edited.
+
+### Update Process
+
+1. Edit alert rules in `jsonnet/custom.libsonnet`
```
Comment (tjungblu) [inline AGENTS.md]: Okay this is a bit more complex and never documented.  Here's just the stuff from the top of my head:
* all files in this directory can be changed, only the lock file and the vendor folder are off limits 
* custom.libsonnet can be used for overrides of upstream alerts or addition of new alerts (downstream)
* If we're changing upstream alerts, e.g for different thresholds, we need to exclude them in main.jsonnet - note that the exclusion must be in sorted order, otherwise the comparison does not work (it's doing a sort dedupe). An agent should verify that the alert is reflected correctly in the generated output. 
* custom dashboards/panels we add should be the own dedicated file and be merge-appended into dashboard.jsonnet
* in general we avoid adding new alerts, because they are a nightmare to maintain in CI - so we very very very rarely add one


When changing any alert, ensure that we update the upstream alerts first. We may want to spawn an agent to keep them rebased at certain invervals in the future

File: AGENTS.md [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+        "EtcdMemberIPMigratorDegraded",  // removed in 4.5
+    },
+    operatorClient,
+    controllerContext.EventRecorder,
+)
+```
+
+See:
+- https://github.com/openshift/cluster-etcd-operator/pull/517
+- https://bugzilla.redhat.com/show_bug.cgi?id=1851351
```
Comment (tjungblu) [inline AGENTS.md]: I would kinda leave this section out, because we don't really do that anymore. But props to digging out that 4.5 issue lol

File: AGENTS.md [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+
+## What NOT to Do
```
Comment (tjungblu) [inline AGENTS.md]: most of those below seem redundant with the critical rules

File: AGENTS.md [hunk 17]
```diff
@@ -0,0 +1,258 @@
...
+7. **Don't remove controllers without cleanup logic** - stale conditions block upgrades
+8. **Don't edit go.sum** - update dependencies in go.mod and then run `go mod tidy && go mod vendor`
+
+## Common Mistakes to Avoid
+
+1. **Assuming 3-node quorum everywhere** - respect the bootstrap scaling strategy (see `ceohelpers/bootstrap.go`)
+2. **Not handling bootstrap teardown** - controllers may run during bootstrap with incomplete state
+3. **Ignoring static pod revision mechanics** - follow library-go patterns for updates
+4. **Breaking IP-based fallback** - etcd must work when DNS is down
+5. **Not considering upgrade paths** - resources/conditions from N-1 may still exist
```
Comment (tjungblu) [inline AGENTS.md]: they are not terrible, but I asked claude to scan our commits and see what failures we're mostly prone to. I think this reflects also my own experience:


```  
  Top 5 Recurring Fix Patterns

  1. Bootstrap, Member Management & Quorum (42 commits)

  The single most-fixed area. Recurring themes:

  - Revision stability vs. member removal timing — OCPBUGS-74151 alone has 5 commits across releases. The tension is: do you wait for revision stability before
  removing members, or does that block bootstrap completion? This was fixed, reverted, and re-fixed across 4.13–4.16 (OCPBUGS-14673, 17199), reverted again, then
  re-approached from a different angle.
  - Bootstrap IP ordering — OCPBUGS-55404 (3 commits): non-deterministic IP selection caused inconsistent bootstrap behavior.
  - Quorum safety during bootstrap — OCPBUGS-42809/42810: quorum wasn't always ensured, and leader transfer wasn't happening before member removal.
  - Scale-down flakes — TestAttemptToScaleDown was fixed 3 separate times for the same flaky test.

  Root cause pattern: Distributed state machine logic with many edge cases around when it's safe to add/remove members. Changes in one release branch get
  backported but interact differently with other release-specific code.

  2. Health Checks, Liveness & Readiness (40 commits)

  - Panics in health check timeouts — OCPBUGS-27959 required 5 commits across releases to fix nil pointer panics when health checks timed out.
  - Liveness probe too aggressive — OCPBUGS-7729 (3 commits): timeout/failure thresholds were too low, causing unnecessary pod restarts. A separate attempt to add
  controller health to liveness (OCPBUGS-11683) was reverted entirely.
  - Aliveness check scope creep — OCPBUGS-30873 (3 commits): the CEO aliveness check was triggering on things other than deadlocks.
  - Client timeout configuration — OCPBUGS-30169: added a timeout to etcd client.New, then reverted it because it caused other issues. Pooled client approach
  (OCPBUGS-38573) was also reverted.
  - Health check ordering — OCPBUGS-36462/37820: non-deterministic member health check ordering caused flapping status.

  Root cause pattern: Getting the sensitivity right is hard — too aggressive kills healthy pods, too relaxed misses real failures. Fixes in one direction tend to
  create problems in the other.

  3. Backup & Restore (40 commits)

  - etcdctl caching on backup — OCPBUGS-19052 (9 commits, the single most-backported bug): cluster-backup.sh was caching etcdctl binaries, causing version
  mismatches.
  - Backup schedule/CR handling — OCPBUGS-43687/43676: empty schedule fields in the EtcdBackup CR weren't handled, causing panics or no-ops.
  - Prune path and retention — multiple fixes to the backup-server pruning logic: wrong paths, wrong defaults, wrong retention counts.
  - Snapshot file locking — OCPBUGS-15327 (4 commits): default snapshot counts weren't reset, causing "file already locked" warnings.
  - Debug pod namespace — The backup e2e test debug pod namespace was fixed 3 separate times.

  Root cause pattern: Shell script fragility (cluster-backup.sh, cluster-restore.sh) and incomplete validation of CR fields. The backup server was built
  incrementally and each piece (scheduling, pruning, retention) had its own round of fixes.

  4. Revision Rollout & Static Pod Management (34 commits)

  - revision.json persistence — OCPBUGS-60273 (4 commits): revision.json didn't survive ungraceful shutdowns, causing revision tracking to desync.
  - Bundle revision rollout — ETCD-606: batch bundle rollout was implemented, reverted, and re-landed.
  - Revision stability checks — the same "remove revision stability check from bootstrap completeness" fix (OCPBUGS-14673/17199) was applied, reverted across 3
  release branches, then re-applied. This single issue accounts for ~10 commits including reverts.
  - Race conditions during cert render — OCPBUGS-32217: concurrent cert creation during static pod render caused races.

  Root cause pattern: The static pod revision mechanism is the critical path for config rollout. Getting the balance wrong between "wait for stability" and "don't
  block bootstrap" causes cascading issues. Each release branch has subtly different behavior, making backports risky.

  5. Race Conditions, Ordering & Informer/Cache Issues (29 commits)

  - Missing informer syncs — OCPBUGS-44570 (2 commits): informers weren't fully synced before controllers started making decisions, leading to stale data.
  - Wrong nodelister scope — OCPBUGS-29717/30245 (4 commits): using the general nodelister instead of master-only nodelister caused controllers to see non-master
  nodes, leading to incorrect scaling decisions.
  - Controller inertia — OCPBUGS-38659 (4 commits): controllers reacted too quickly to transient state changes, requiring inertia configuration to dampen flapping
  (bumped from 5 to 10 minutes).
  - Shared informer conflicts — dedicated informers had to be created to avoid conflicts between controllers watching the same resources.
  - Test flakiness — defrag controller race, scale-down test flakes, and general UT flakiness each required dedicated fixes.

  Root cause pattern: Kubernetes informer/lister caching semantics are subtle. Using the wrong lister scope or not waiting for sync is easy to miss in review. The
  inertia issue shows that even correct logic can misbehave when controllers react to intermediate states during rollouts.

  ---
  Notable cross-cutting observation

  19 reverts in this period — a high rate. The most-reverted areas are health check changes (3 reverts), revision stability logic (3 reverts), and bundle rollout
  (1 revert). This suggests that changes in the etcd lifecycle management space are hard to validate fully before merge, and the blast radius of getting them wrong
  is high enough that reverting is the safe default.


```

File: AGENTS.md [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+
+## Security Notes
+
+1. **TLS Certificates** - managed by `etcdcertsigner` and `etcdcertcleaner` controllers
```
Comment (tjungblu) [inline AGENTS.md]: another super annoying part are the TLS ciphers. Maybe you can prompt a section for handling that more specifically. But this is a moving target as well, there are some changes inbound

File: AGENTS.md [hunk 19]
```diff
@@ -0,0 +1,258 @@
...
+2. **Backup Credentials** - stored in secrets, never logged
```
Comment (tjungblu) [inline AGENTS.md]: that's a hallucination

File: AGENTS.md [hunk 20]
```diff
@@ -0,0 +1,258 @@
...
+3. **Escape Hatches** - documented in docs/overview/overview.md (Secret Escape Hatches section)
```
Comment (tjungblu) [inline AGENTS.md]: not so secret anymore, is it :) 

I think the better way to deal with such a thing would be more of a feature gate. We didn't have that back then and because of severe stability problems, hence the escape, we added a flag configmap.

Maybe we can devote a section for it, that may be relevant for feature development in this repo

File: AGENTS.md [hunk 21]
```diff
@@ -0,0 +1,258 @@
...
+4. **Fencing Scripts** - in bindata/etcd/ (shell scripts) and bindata/tnfdeployment/ (deployment manifests), reviewed for injection vulnerabilities
```
Comment (tjungblu) [inline AGENTS.md]: @jaypoulz would be that relevant for your teams day2day agentic workflow?

File: AGENTS.md [hunk 22]
```diff
@@ -0,0 +1,258 @@
...
+
```
Comment (tjungblu) [inline AGENTS.md:265]: just to confirm my own comments, asked claude to summarize our security vulnerabilities from the commit history:

``` 
  Security & CVE Fix Patterns

  Overview

  ~50 security-related commits since 2023, falling into 5 distinct categories. The dominant pattern is transitive dependency CVEs requiring backports across many
  release branches.

  ---
  1. Transitive Go Dependency CVEs (30+ commits)

  This is the overwhelming majority of security work. Three CVEs account for the bulk:

  ┌────────────────┬───────────────────────────────────┬───────────────────────────────┬────────────────┬─────────────────────┐
  │      CVE       │              Library              │             Issue             │    Commits     │ Branches Backported │
  ├────────────────┼───────────────────────────────────┼───────────────────────────────┼────────────────┼─────────────────────┤
  │ CVE-2025-30204 │ golang-jwt/jwt/v4 (4.4.2 → 4.5.2) │ JWT parsing vulnerability     │ 8              │ 14 (4.12–5.1)       │
  ├────────────────┼───────────────────────────────────┼───────────────────────────────┼────────────────┼─────────────────────┤
  │ CVE-2023-39325 │ golang.org/x/net/http2            │ HTTP/2 rapid reset DoS        │ 6 (1 reverted) │ 15 (4.11–5.1)       │
  ├────────────────┼───────────────────────────────────┼───────────────────────────────┼────────────────┼─────────────────────┤
  │ CVE-2022-41717 │ golang.org/x/net (via library-go) │ HTTP/2 memory exhaustion      │ 2              │ 2                   │
  ├────────────────┼───────────────────────────────────┼───────────────────────────────┼────────────────┼─────────────────────┤
  │ CVE-2025-65637 │ sirupsen/logrus (1.8.1 → 1.9.3)   │ Logging library vulnerability │ 1              │ 1 (so far)          │
  └────────────────┴───────────────────────────────────┴───────────────────────────────┴────────────────┴─────────────────────┘

  The pain: These are not bugs in this repo's code — they're bumps to indirect dependencies. But each bump must be backported to every supported release branch
  independently, because each branch pins different versions. CVE-2025-30204 alone produced 8 commits across 14 branches. CVE-2023-39325 hit 15 branches and even
  had to be reverted on 4.11 because the dependency bump broke the older Go version.

  In addition, golang.org/x/net was updated 5 more times in April 2024 (OCPBUGS-319xx series) across 5 release branches — likely for another x/net advisory that
  wasn't given a CVE in the commit messages.

  Root cause: No automated dependency update mechanism across release branches. Each branch is a manual effort with its own go.mod, and older branches have tighter
  Go version constraints that can make a straightforward bump impossible.

  ---
  2. TLS/Cipher Suite Configuration (10 commits, 1 revert)

  A multi-year saga:

  1. Dec 2022: Hardcoded safe TLS 1.2/1.3 cipher suites in etcd and CEO
  2. Aug 2024 (OCPBUGS-38341): Switched from hardcoded ciphers to environment variable — hardcoding broke when the platform's TLS policy changed
  3. May 2025 (OCPBUGS-56834): etcd pods wouldn't start when TLS min version was set to 1.3 — the cipher suite list and the readyz probe didn't support TLS
  1.3-only mode. Required touching 6 files including the pod template, readyz handler, and env var controller
  4. Oct 2025 (ensure etcd starts with TLS 1.3): Another fix for TLS 1.3 compatibility
  5. Sep 2025: "use secure cipher suites" committed
  6. Mar 2026: Reverted — the secure cipher suite change broke something

  Root cause: The cipher suite / TLS version configuration sits at the intersection of three systems: the platform's crypto policy (apiserver TLS profile), etcd's
  own TLS config, and the operator's readyz/liveness probes. Changes to one system expose incompatibilities in the others. The pattern of commit-then-revert shows
  that testing TLS configuration across all supported profiles is difficult.

  ---
  3. Missing Container Auth Credentials (6 commits)

  OCPBUGS-18357 (prepend authfile to podman create) was backported 6 times across release branches. The cluster-backup.sh script was running podman create without
  the --authfile flag, meaning on nodes where registry auth is required, the backup container couldn't pull images.

  Root cause: Shell scripts don't have the same level of review scrutiny as Go code, and the authfile flag is easy to forget since podman works without it in many
  dev/test environments that have pre-cached images.

  ---
  4. Network Policy Hardening (4 commits, 2025–2026)

  A relatively recent addition:

  - CNTRLPLANE-2663: Added network policies to the operator
  - CNTRLPLANE-2683: Added network policies to openshift-etcd pods
  - TNF jobs added to network policy exceptions
  - E2e tests added for network policy

  Root cause: The etcd namespace previously ran without network policies, relying on the cluster's default-deny being absent. When default-deny became standard,
  etcd communication broke unless explicit policies existed. This was a gap rather than a bug — the policies should have existed from the beginning.

  ---
  5. Operator Restart on TLS Config Changes (1 commit, late addition)

  a79ec721a: Restart operator when TLS config changes — the operator wasn't watching for changes to the platform TLS configuration. When an admin changed the
  crypto policy, the operator kept running with the old TLS settings until it was manually restarted.

  Root cause: The TLS config was treated as immutable at startup rather than being watched as a dynamic config source.

  ---
  Cross-cutting observations

  - Backport tax is the dominant cost. The actual fix for each CVE is typically a one-line go.mod bump. The work is doing it 8–15 times across release branches,
  each with its own dependency graph. CVE-2023-39325 showed this can actively backfire — the 4.11 backport was reverted because it was incompatible with that
  branch's Go version.
  - TLS is the only area where this repo's own code has recurring security issues. Everything else is dependency management. The cipher suite saga (4 years, still
  getting reverted) suggests the configuration surface is too complex for the testing that exists.
  - Shell scripts are a blind spot. The authfile fix is a security issue that wouldn't have been caught by Go-level tooling (linting, go vet, etc.). The
  backup/restore scripts interact with container runtimes and need their own security review process.



```

File: AGENTS.md (hunk 21 above)
Comment (tjungblu) [reply AGENTS.md]: I also think, because it's gotten so big in the meantime, we should devote an entire TNF section here too...

File: CONTRIBUTING.md [hunk 23]
```diff
@@ -0,0 +1,173 @@
...
+### Interacting with reviewers
+
+When interacting with reviewers/approvers:
+
+- Be professional.
+- Be respectful of differing opinions, viewpoints, and experiences.
+- Gracefully give and receive constructive feedback.
+- Focus on what is best for the product/organization, not just us as individuals.
+
+A special note on the usage of AI - to respect the time of those that are reviewing your contribution, please do not use AI to respond to review comments.
```
Comment (tjungblu) [inline CONTRIBUTING.md:173]: it's okay, we don't have to be respectful around here 🤣

File: ARCHITECTURE.md [hunk 24]
```diff
@@ -0,0 +1,315 @@
...
+                           ▼
+            ┌──────────────────────────────┐
+            │  etcd static pods (per node) │
+            │  ┌─────────┐  ┌─────────┐    │
+            │  │  etcd   │  │  etcd-  │    │
+            │  │ :2379   │  │ metrics │    │
+            │  │ :2380   │  │ :9979   │    │
+            │  └─────────┘  └─────────┘    │
+            └──────────────────────────────┘
+```
```
Comment (tjungblu) [inline ARCHITECTURE.md:29]: can we maybe make this a mermaid chart instead?

File: ARCHITECTURE.md [hunk 25]
```diff
@@ -0,0 +1,315 @@
...
+
+## Controllers
+
+All controllers are initialized in `pkg/operator/starter.go` `RunOperator()`. They follow the library-go factory pattern, wrapping their sync logic with `health.NewDefaultCheckingSyncWrapper()` for liveness monitoring via a centralized `MultiAlivenessChecker`.
```
Comment (tjungblu) [inline ARCHITECTURE.md]: maybe to add why we're doing this:
* our operator was deadlocked very frequently
* we couldn't figure it out for a long time, even with thread dumps
* we accidentally caught one stuck in grpc in etcdclient, but there are some other places for sure

So this is band-aid to restart the operator in order for it to continue processing. If you're old enough, this is kinda like the daily Tomcat/Apache restart ;-)

File: ARCHITECTURE.md [hunk 26]
```diff
@@ -0,0 +1,315 @@
...
+
+### Cluster Membership Controllers
```
Comment (tjungblu) [inline ARCHITECTURE.md:42]: I personally don't mind to have a shorter technical summary of every controller - but it has nothing to do with architecture. 

To me the architectural considerations in this operator are the following:
* quorum safety first - we favor etcd availability ALWAYS and on any SKU, anything that causes downtime must be avoided at all costs
* one concern per controller (needs a bit more nuance, but let's leave this here for now)
* always use informers and cached clients, avoid directly hitting APIs and etcd, we want a low resource control plane footprint
* always refactor into reusable helpers and interfaces as building blocks
* must-gather readability - debugging CI and customer cases is difficult. We want to be generally verbose in logging, only use status conditions and events when you want to communicate actionable feedback to the admin or the CI timeline

File: AGENTS.md (hunk 9 above)
Comment (tjungblu) [reply AGENTS.md:110]: oh and more general openshift knowledge won't hurt: we support all control plane sizes now from 1 node to 5 nodes.

Plus microshift, which actually has some special cases that we need to be aware of here. @dusk125 I think that was mostly around the embedding into the binary, so we can't ever be ahead in go version and we have to be careful about the dependencies.

File: AGENTS.md (hunk 10 above)
Comment (tjungblu) [reply AGENTS.md]: I saw this is also covered below more succinct as "no circular dependency on coreDNS"

File: AGENTS.md [hunk 27]
```diff
@@ -0,0 +1,258 @@
...
+1. **DNS Independence** - The `EtcdEndpointsController` must never depend on DNS (directly or transitively) as it maintains the IP-based ConfigMap that DNS depends on
+2. **CVO Resource Management** - Resources in `manifests/` are managed by CVO; removal requires special handling (see [CVO Resource Removal](#cvo-resource-removal))
+3. **Controller Removal** - Removing a controller requires cleaning up stale status conditions to prevent upgrade blockers
+4. **Static Pod Revisions** - Follow library-go static pod revision pattern; don't invent custom update logic
+5. **Namespace** - Primary namespace is `openshift-etcd`; target namespace constant is in `pkg/operator/operatorclient`
+
+## Build and Test
+
+### Common Commands
```
Comment (tjungblu) [inline AGENTS.md:128]: one last open ended request for you - sorry for all the comments :) 

I always use agents to debug live systems for me, so it would be quite useful to tell it about https://github.com/openshift/cluster-etcd-operator/blob/main/docs/profiling_and_debugging.md

and https://github.com/openshift/cluster-etcd-operator/blob/main/hack/unmanage.sh

basically to show how to:
* run CEO locally, connect to remote cluster (means ./hack/unmanage.sh && build && run) 
* run a local build of CEO inside the cluster (like the above, just with pushing and updating the image)

But I can add this later if you're running out of time.

Comment (bhperry) [reply AGENTS.md:128]: Appreciate the thorough review. Happy to add those in.

File: ARCHITECTURE.md (hunk 26 above)
Comment (bhperry) [reply ARCHITECTURE.md:42]: Agreed, I wanted to avoid cluttering the main AGENTS.md context too much but still thought a description of the controllers could be useful. ARCHITECTURE.md seemed like a reasonable place for that, but we could also make a separate docs/controllers.md file linked from architecture if you would prefer.

These bullet points are helpful, will add them.

File: AGENTS.md (hunk 27 above)
Comment (bhperry) [reply AGENTS.md:128]: Added in a rough overview of running CEO locally. Was able to get it running with this, but there were some errors still, not sure if I need to also fetch all the certs locally or if I'm missing something else.

Comment (tjungblu) [reply AGENTS.md:128]: Thank you! I wanted to also say that we can do it iteratively and I can write a few sections over the course of the next sprint as well. 

On the error topic: yes, that's a bit annoying because you can't reach the etcd members locally, so it may not just be a matter of certificates, but also port forwarding and re-configuring the etcdclient accordingly. Then the certs may not match anymore? 

My own rule of thumb is: any controller that doesn't need the etcdclient is debuggable from local, everything else needs to be deployed in cluster. But maybe there is a better way to proxy into the host network nowadays.

File: ARCHITECTURE.md (hunk 26 above)
Comment (tjungblu) [reply ARCHITECTURE.md:42]: Architecture.md is a good place for it, no doubts. Thanks for adding the bullets.

File: AGENTS.md (hunk 27 above)
Comment (bhperry) [reply AGENTS.md:128]: Agreed. This is just the starting point.

It would be really nice to be able to dev locally against a remote cluster with etcd proxied. Something to think on for later.

Comment (tjungblu) [reply AGENTS.md:128]: yeah I can do that. Do you consider this ready for a v1? I can label it and we can override the failing job.

File: AGENTS.md (hunk 9 above)
Comment (dusk125) [reply AGENTS.md:110]: Yeah that's correct, and that fact `so we can't ever be ahead in go version and we have to be careful about the dependencies.` gently reared its head again in the intern project we're currently working on. 

This is _currently_ true, but may, in the (hopefully short) future, it won't be - ART is looking into this so we could be "ahead"
